### PR TITLE
feat(finemapping): ingest finngen r10 finemapping w/ airflow

### PR DESCRIPTION
**Pull request by xyg123**
Working commit that generated the finngen finemapping results from finngen r10.

### CODE REVIEW

**Comment by d0choa** [comment]
@xyg123 let us know when you are done with the comments from #394

**Comment by xyg123** [comment]
> @xyg123 let us know when you are done with the comments from #394
Done!

File: src/gentropy/datasource/finngen/finngen_finemapping.py [hunk 1]
```diff
@@ -0,0 +1,286 @@
+# pylint: disable=unsubscriptable-object
```
Comment (d0choa) [inline src/gentropy/datasource/finngen/finngen_finemapping.py:1]: I don't think this line is necessary.

File: src/gentropy/datasource/finngen/finngen_finemapping.py [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+"""Datasource ingestion: FinnGen Finemapping results (SuSIE) to studyLocus object."""
```
Comment (d0choa) [inline src/gentropy/datasource/finngen/finngen_finemapping.py:2]: For consistency, I would rename the module to `/finngen/finemapping.py`. You are missing this part in the documentation `/docs/python_api/datasources/finemapping.md`

File: config/datasets/ot_gcp.yaml [hunk 3]
```diff
@@ -40,19 +40,24 @@ target_index: ${datasets.release_folder}/targets # OTP 23.12 data
 
 gene_interactions: ${datasets.release_folder}/interaction # OTP 23.12 data
 eqtl_catalogue_paths_imported: ${datasets.inputs}/preprocess/eqtl_catalogue/tabix_ftp_paths_imported.tsv
+finngen_finemapping_results_url: ${datasets.inputs}/Finngen_susie_finemapping_r10/full
```
Comment (d0choa) [inline config/datasets/ot_gcp.yaml:43]: I would avoid using `url` and use `path` or something similar instead. `url` refers to a location in the web which is not the case.

File: src/gentropy/finngen_finemapping_ingestion.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        Args:
+            session (Session): Session object.
+            finngen_finemapping_results_url (str): URL to the FinnGen SuSIE finemapping results.
+            finngen_finemapping_summaries_url (str): FinnGen SuSIE summaries for CS filters(LBF>2).
+            finngen_release_prefix (str): Release prefix for FinnGen.
+            finngen_finemapping_out (str): Output path for the finemapping results in StudyLocus format.
+        """
+        # Read finemapping outputs from the URL.
+
+        # finngen_finemapping_results = glob.glob(finngen_finemapping_results_url)
```
Comment (d0choa) [inline src/gentropy/finngen_finemapping_ingestion.py:35]: remove commented code for final version.

**Review by d0choa — COMMENTED** [review]
This is almost done. Please have a look at the comments. The required changes are minimal but some are important.

@ireneisdoomed could you help confirm the paths are aligned with the latest data structures?

I also discussed with @addramir whether there is a case to abstract the reading of SuSie finemapping results since we would like to also harmonise other SuSie outputs (e.g. eQTL catalogue). After looking at the catalogue's SuSie outputs, we concluded that although there are some commonalities in the data structure, the schemas diverge enough to not justify a common parser.

File: src/gentropy/datasource/finngen/finngen_finemapping.py (hunk 1 above)
Comment (xyg123) [reply src/gentropy/datasource/finngen/finngen_finemapping.py]: Without this it throws me an error for type checking where I have list[str]

**Comment by xyg123** [comment]
> This is almost done. Please have a look at the comments. The required changes are minimal but some are important.
> 
> @ireneisdoomed could you help confirm the paths are aligned with the latest data structures?
> 
> I also discussed with @addramir whether there is a case to abstract the reading of SuSie finemapping results since we would like to also harmonise other SuSie outputs (e.g. eQTL catalogue). After looking at the catalogue's SuSie outputs, we concluded that although there are some commonalities in the data structure, the schemas diverge enough to not justify a common parser.

Thank you for the comments!  I've applied all of them.

File: src/gentropy/datasource/finngen/finemapping.py [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+
+    - Study ID in the special format (e.g. FINNGEN_R10_*)
+    - Credible set specific finemapping statistics (e.g. LogBayesFactors, Alphas/Posterior)
+    - Additional credible set level BayesFactor filtering is applied (LBF > 2)
+    - StudyLocusId is annotated for each credible set.
+
+    Finemapping method is populated as a constant ("SuSIE").
+    """
+
+    raw_schema: t.StructType = StructType(
```
Comment (ireneisdoomed) [inline src/gentropy/datasource/finngen/finemapping.py:32]: The schema can live in `assets/schemas`

File: config/step/ot_finngen_finemapping_ingestion.yaml [hunk 6]
```diff
@@ -0,0 +1,7 @@
+defaults:
+  - finngen_finemapping_ingestion
+
+finngen_finemapping_results_url: ${datasets.finngen_finemapping_results_url}
```
Comment (ireneisdoomed) [inline config/step/ot_finngen_finemapping_ingestion.yaml:4]: I think this is outdated
```suggestion
finngen_finemapping_results_url: ${datasets.finngen_finemapping_results_path}
```

File: config/step/ot_finngen_finemapping_ingestion.yaml [hunk 7]
```diff
@@ -0,0 +1,7 @@
...
+finngen_finemapping_summaries_url: ${datasets.finngen_finemapping_summaries_url}
```
Comment (ireneisdoomed) [inline config/step/ot_finngen_finemapping_ingestion.yaml:5]: I think this is outdated
```suggestion
finngen_finemapping_summaries_url: ${datasets.finngen_finemapping_summaries_path}
```

**Review by ireneisdoomed — COMMENTED** [review]
Thank you for adding my suggestions @xyg123!

**Comment by d0choa** [comment]
IMO only one bit is missing. Both `finngen_finemapping_results_path` and `finngen_finemapping_summaries_path` point to non-existent datasets in our bucket. We should be pulling this data directly from Finngen buckets without the need to copy any data over to our buckets (as we do for the summary statistics).

**Comment by ireneisdoomed** [comment]
> The problem is that the finngen buckets stores these files in .bgz compressed format, which spark just cannot seem to read. To process this release I had to download them, unzip and re-zip them to .gz

@xyg123 Spark can read `bgz` compressed files, but for some reason it is the recognition of the format what is failing. If you rename the file from `bgz` to `gz`, data can be loaded:
```
spark.read.csv("finngen_R10_AB1_OTHER_INFECTIONS.SUSIE.snp.gz", sep="\t", header=True).show(1, False, True)
-RECORD 0---------------------------------------
 trait               | AB1_OTHER_INFECTIONS     
 region              | chr1:211076894-214076894 
 v                   | 1:211077031:A:G          
 rsid                | chr1_211077031_A_G       
 chromosome          | chr1                     
 position            | 211077031 
```
However interestingly enough, specifying the format doesn't help. Maybe someone from @opentargets/be-team knows more about it?
```
path = "gs://finngen-public-data-r10/finemap/full/susie/finngen_R10_AB1_OTHER_INFECTIONS.SUSIE.snp.bgz"
spark.read.option("compression", "bzip2").csv(path, sep="\t", header=True).show(1)
```

In any case, if you can't ingest the data directly, you can orchestrate this operation (renaming) in your DAG using an Airflow operator to prepare the data before ingestion with Spark. You can check `src/airflow/dags/gwas_catalog_harmonisation.py` for inspiration. As operators you could use one to copy [to your local system](https://airflow.apache.org/docs/apache-airflow-providers-google/stable/operators/transfer/gcs_to_local.html) or [to another bucket](https://airflow.apache.org/docs/apache-airflow-providers-google/stable/_api/airflow/providers/google/cloud/transfers/gcs_to_gcs/index.html).

**Comment by addramir** [comment]
@ireneisdoomed  As discussed with @d0choa  - for this PR we could just store the .gz data in the "genetics_etl_python_playground" bucket and create the separate PR for the orchestration (downloading and converting the data) later. Please confirm that this is an acceptable way to proceed?

**Comment by ireneisdoomed** [comment]
@addramir Absolutely, this PR covers already a lot of edges

**Comment by d0choa** [comment]
Let's try to make it run. @xyg123  can you just make sure that  `finngen_finemapping_results_path` and `finngen_finemapping_summaries_path` point to valid paths containing the necessary inputs to run this step for Finngen R10? As you said, we can make additional work in separate PRs

**Review by d0choa — APPROVED** [review]
Thanks for all the work. We should make sure all the layers integrate well for the next data release. There are a lot of moving pieces in this PR
When you are ready @xyg123

**Pull request merged**